src/PowerShellEditorServices.Host/CodeLens/ReferencesCodeLensProvider.cs

```diff
@@ -14,87 +15,120 @@
         public async Task<CodeLens> ResolveCodeLensAsync(
             CodeLens codeLens,
             CancellationToken cancellationToken)
         {
-            ScriptFile[] references =
-                editorSession.Workspace.ExpandScriptReferences(
-                    codeLens.File);
-
-            var foundSymbol =
-                this.editorSession.LanguageService.FindFunctionDefinitionAtLocation(
-                    codeLens.File,
-                    codeLens.ScriptExtent.StartLineNumber,
-                    codeLens.ScriptExtent.StartColumnNumber);
-
-            FindReferencesResult referencesResult =
-                await editorSession.LanguageService.FindReferencesOfSymbol(
-                    foundSymbol,
-                    references,
-                    editorSession.Workspace);
-
-            Location[] referenceLocations =
-                referencesResult == null
-                    ? new Location[0]
-                    : referencesResult
-                        .FoundReferences
-                        .Where(r => NotReferenceDefinition(foundSymbol, r))
-                        .Select(
-                            r => new Location
-                            {
-                                Uri = GetFileUri(r.FilePath),
-                                Range = r.ScriptRegion.ToRange()
-                            })
-                        .ToArray();
+            ScriptFile[] references = _editorSession.Workspace.ExpandScriptReferences(
+                codeLens.File);
 
-            return
-                new CodeLens(
-                    codeLens,
-                    new ClientCommand(
-                        "editor.action.showReferences",
-                        referenceLocations.Length == 1
-                            ? "1 reference"
-                            : $"{referenceLocations.Length} references",
-                        new object[]
-                        {
-                            codeLens.File.ClientFilePath,
-                            codeLens.ScriptExtent.ToRange().Start,
-                            referenceLocations,
-                        }
+            SymbolReference foundSymbol = _editorSession.LanguageService.FindFunctionDefinitionAtLocation(
+                codeLens.File,
+                codeLens.ScriptExtent.StartLineNumber,
+                codeLens.ScriptExtent.StartColumnNumber);
+
+            FindReferencesResult referencesResult = await _editorSession.LanguageService.FindReferencesOfSymbol(
+                foundSymbol,
+                references,
+                _editorSession.Workspace);
+
+            Location[] referenceLocations;
+            if (referencesResult == null)
+            {
+                referenceLocations = Array.Empty<Location>();
+            }
+            else
+            {
+                var acc = new List<Location>();
+                foreach (SymbolReference foundReference in referencesResult.FoundReferences)
```

Comment:
I feel like all these loops could be a little more DRY.

Reply:
As in each particular loop? Or the fact we are doing loops like this in each call?

They were decidedly pithier as LINQ expressions, but that meant inefficiencies. The problem is that the logic inside each loop is different, so we either write out the loop or use some kind of function pointer or closure.

But I guess to address the DRY need, we should:
- Identify what the repetition you're seeing is, and
- Work out what the best refactoring is that preserves efficiency

Reply:
I thought the foreach's were all doing similar things but they're actually quite different :) we're good